contracts/message/apps/RFQ.sol

```diff
@@ -0,0 +1,266 @@
+        bytes32 quoteHash = getQuoteHash(_quote);
+        receiveMsgAndCheckHash(_message, _route, _sigs, _signers, _powers, quoteHash);
+        require(quotes[quoteHash] == true, "incorrect agreement hash");
+        delete quotes[quoteHash];
+        delete unconsumedMsg[quoteHash];
+        address receiver = (_quote.refundTo == address(0)) ? _quote.sender : _quote.refundTo;
+        IERC20(_quote.srcToken).safeTransfer(receiver, _quote.srcAmount);
+        emit Refunded(quoteHash, receiver, _quote.srcToken, _quote.srcAmount);
+    }
+
+    function executeMessage(
+        address _sender,
+        uint64 _srcChainId,
+        bytes calldata _message,
+        address // executor
+    ) external payable override onlyMessageBus returns (ExecutionStatus) {
+        address expectedSender = remoteRfqContracts[_srcChainId];
+        if (expectedSender != _sender) {
+            return ExecutionStatus.Retry;
+        }
+
+        bytes32 quoteHash = abi.decode(_message, (bytes32));
+        unconsumedMsg[quoteHash] = true;
+
+        emit MessageReceived(quoteHash);
+        return ExecutionStatus.Success;
+    }
+
+    //=========================== helper functions ==========================
+
+    function getQuoteHash(Quote calldata _quote) public pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    _quote.srcChainId,
+                    _quote.srcToken,
+                    _quote.srcAmount,
+                    _quote.dstChainId,
+                    _quote.dstToken,
+                    _quote.dstAmount,
+                    _quote.deadline,
+                    _quote.nonce,
+                    _quote.sender,
+                    _quote.receiver,
+                    _quote.refundTo,
+                    _quote.liquidityProvider
+                )
+            );
+    }
+
+    function getRFQFee(uint64 _dstChainId, uint256 _amount) public view returns (uint256) {
+        uint32 feePerc = feePercOverride[_dstChainId];
+        if (feePerc == 0) {
+            feePerc = feePercGlobal;
+        }
+        return (_amount * feePerc) / 1e6;
+    }
+
+    function receiveMsgAndCheckHash(
+        bytes calldata _message,
+        MsgDataTypes.RouteInfo calldata _route,
+        bytes[] calldata _sigs,
+        address[] calldata _signers,
+        uint256[] calldata _powers,
+        bytes32 _quoteHash
+    ) private {
+        bytes32 expectedQuoteHash = abi.decode(_message, (bytes32));
+        require(_quoteHash == expectedQuoteHash, "mismatch agreement hash");
+        if (unconsumedMsg[_quoteHash] == false) {
+            IMessageBus(messageBus).executeMessage(_message, _route, _sigs, _signers, _powers);
+        }
+        assert(unconsumedMsg[_quoteHash] == true);
+    }
+
+    //=========================== admin operations ==========================
+
+    function collectFee(address _token, uint256 _amount) external onlyOwner {
```

Comment:
Don't need `onlyOwner` if fees can only go to the treasury.

Reply:
The contract should record uncollected fees. Then change this function to 
`function collectFee(address _token) external {...}`